### PR TITLE
Mobile app api authentication exemptions

### PR DESCRIPTION
**Pull request by 123MwanjeMike**
**_WHAT DOES THIS PR DO?_**
- Adds api authentication at the load balancer

**_WHAT ISSUES ARE RELATED TO THIS PR?_**
- Jira cards
    - [AN-160](https://airqoteam.atlassian.net/browse/AN-160?atlOrigin=eyJpIjoiNjZlMTViMzM4YzRmNDc1Zjk3OTQwNTMzZmYwMmU2NjIiLCJwIjoiaiJ9)

**_HOW DO I TEST OUT THIS PR?_**
 - Apply the staging ingess file
 - Attempt to get the devices 
 - You should get an "Unauthorized" response body
    image
 - Repeat the process but this time with an authentication token in the Authorization header
 - You should get the list of devices.
   image

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 
- Any endpoint

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
- n/a

**_ARE THERE ANY RELATED PRs?_**
- #1131

### CODE REVIEW

**Comment by 123MwanjeMike** [comment]
Next, we need the auth-service to automatically authorise;
- [ ] the endpoints that do not need authentication e.g login, register, forgot password, and home(/) 
- [ ] other internal micro services and automatically renew the respective tokens.

cc @Baalmart

File: k8s/nginx-ingress/prod-api-ingress-resource.yaml [hunk 1]
```diff
@@ -3,6 +3,24 @@ kind: ConfigMap
 data:
   proxy-body-size:  "200m"
   client-max-body-size: "200m"
+  location-snippets: "
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/prod-api-ingress-resource.yaml:6]: Adding location and server snippets via the configmap

File: k8s/nginx-ingress/prod-api-ingress-resource.yaml [hunk 2]
```diff
@@ -3,6 +3,24 @@ kind: ConfigMap
...
+    auth_request /auth;
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/prod-api-ingress-resource.yaml:7]: All endpoints require authorization by default

File: k8s/nginx-ingress/prod-api-ingress-resource.yaml [hunk 3]
```diff
@@ -15,56 +33,6 @@ metadata:
   namespace: production
   annotations:
     kubernetes.io/ingress.class: "nginx" 
-    nginx.ingress.kubernetes.io/rewrite-target: /
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/prod-api-ingress-resource.yaml:18]: Removing annotation configurations as they were no longer working

File: k8s/nginx-ingress/stage-platform-ingress-resource.yaml [hunk 4]
```diff
@@ -3,6 +3,32 @@ kind: ConfigMap
...
+    error_page 404 = @page_not_found;
+    error_page 403 = @access_denied;
+    error_page 401 = @unauthorized;"
+  server-snippets: "
+    location @internal_server_error { return 500 'Something went wrong'; }
+    location @page_not_found { return 404 'Page not found'; }
+    location @access_denied { return 403 'Access denied'; }
+    location @unauthorized { return 401 'Unauthorized'; }
+
+    location = / { auth_request off; return 202; }
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/stage-platform-ingress-resource.yaml:18]: Return 202 for endpoints that do  not need authorization. This is as we await the auth service to automatically handle these scenarios

File: k8s/nginx-ingress/prod-api-ingress-resource.yaml [hunk 5]
```diff
@@ -3,6 +3,33 @@ kind: ConfigMap
...
+    error_page 401 = @unauthorized;"
+  server-snippets: "
+    location @internal_server_error { return 500 'Something went wrong!'; }
+    location @page_not_found { return 404 'Page not found :('; }
+    location @access_denied { return 403 'Access denied :|'; }
+    location @unauthorized { return 401 'Unauthorized'; }
+
+    location = /auth { 
+      if ($request_uri = '/') { return 200; }
+      if ($request_uri ~ '/docs') { return 200; }
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/prod-api-ingress-resource.yaml:20]: Publicly availing the docs

File: k8s/nginx-ingress/prod-platform-ingress-resource.yaml [hunk 6]
```diff
@@ -3,6 +3,33 @@ kind: ConfigMap
...
+    location @access_denied { return 403 'Access denied :|'; }
+    location @unauthorized { return 401 'Unauthorized'; }
+
+    location = /auth {
+      if ($request_uri = '/') { return 200; }
+      if ($request_uri = '/login') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/loginUser)|(/api/v2/users/loginUser)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/forgotPassword)|(/api/v2/users/forgotPassword)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/candidates/register)|(/api/v2/users/candidates/register)') { return 200; }
+      if ($request_uri ^ '/static/') { return 200; }
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/prod-platform-ingress-resource.yaml:24]: Publicly availing the Netmanager static files

File: k8s/nginx-ingress/prod-api-ingress-resource.yaml (hunk 5 above)
Comment (Baalmart) [reply k8s/nginx-ingress/prod-api-ingress-resource.yaml:20]: Awesome, thanks @123MwanjeMike

File: k8s/nginx-ingress/prod-platform-ingress-resource.yaml (hunk 6 above)
Comment (Baalmart) [reply k8s/nginx-ingress/prod-platform-ingress-resource.yaml:24]: Brilliant! thanks @123MwanjeMike

**Review by Baalmart — COMMENTED** [review]
Brilliant work, thanks @123MwanjeMike ...how does this look like from our tentative roadmap?
**What/when is the right time for this PR to be merged?**
Thanks again!

**Comment by 123MwanjeMike** [comment]
> Brilliant work, thanks @123MwanjeMike ...how does this look like from our tentative roadmap?
> **What/when is the right time for this PR to be merged?**
> Thanks again!

This can be merged after we have implemented strategies to allow authentication in our frontends

File: .github/workflows/deploy-apis-to-production.yml [hunk 7]
```diff
@@ -1574,5 +1573,6 @@ jobs:
 
       - name: Apply ingress changes
         run: |
+          kubectl apply -f k8s/nginx-ingress/prod-nginx-cm.yaml
```
Comment (123MwanjeMike) [inline .github/workflows/deploy-apis-to-production.yml:1576]: [add configmap updates in nginx workflow](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/8fda839b6f4d0fe0a98956dbc1b12113e6389934)

File: k8s/nginx-ingress/stage-api-ingress-resource.yaml [hunk 8]
```diff
@@ -0,0 +1,180 @@
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: stage-api-ingress-resource
+  namespace: staging
+  annotations:
+    kubernetes.io/ingress.class: nginx
+spec:
+  rules:    
+    - host: staging-api.airqo.net
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/stage-api-ingress-resource.yaml:10]: [staging api ingress](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/7073b3303a3d4aff78bb75a35c1ee5bef1d6b08c)

File: k8s/nginx-ingress/prod-api-ingress-resource.yaml [hunk 9]
```diff
@@ -1,70 +1,10 @@
-apiVersion: v1
-kind: ConfigMap
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/prod-api-ingress-resource.yaml:2]: [seperate configmap from ingress](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/cb3700f51c49bb534a441aa78fc816c12d065ec0)

File: k8s/nginx-ingress/global-configs/stage-nginx-config.yaml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    error_page 403 = @access_denied;
+    error_page 401 = @unauthorized;"
+  server-snippets: "
+    location @internal_server_error { return 500 '{\"success\":false, \"message\":\"Something went wrong!\", \"error\":\"Internal Server Error\"}'; }
+    location @page_not_found { return 404 '{\"success\":false, \"message\":\"Resource not found :(\", \"error\":\"Not Found\"}'; }
+    location @access_denied { return 403 '{\"success\":false, \"message\":\"Access denied :|\", \"error\":\"Insufficient Permissions\"}'; }
+    location @unauthorized { return 401 '{\"success\":false, \"message\":\"Invalid authentication\", \"error\":\"Unauthorized.\"}'; }
+
+    location = /auth {      
+      if ($request_method = 'OPTIONS') { return 204; }
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/global-configs/stage-nginx-config.yaml:22]: [Authenticate 'OPTIONS' method for staging](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/7d978b4928ce9b1221d8336ebf5eca901c0189a6)

File: k8s/nginx-ingress/global-configs/stage-nginx-config.yaml [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    auth_request /auth;
+    error_page 500 = @internal_server_error;
+    error_page 404 = @page_not_found;
+    error_page 403 = @access_denied;
+    error_page 401 = @unauthorized;"
+  server-snippets: "
+    location @internal_server_error { return 500 '{\"success\":false, \"message\":\"Something went wrong!\", \"error\":\"Internal Server Error\"}'; }
+    location @page_not_found { return 404 '{\"success\":false, \"message\":\"Resource not found :(\", \"error\":\"Not Found\"}'; }
+    location @access_denied { return 403 '{\"success\":false, \"message\":\"Access denied :|\", \"error\":\"Insufficient Permissions\"}'; }
+    location @unauthorized { return 401 '{\"success\":false, \"message\":\"Invalid authentication\", \"error\":\"Unauthorized.\"}'; }
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/global-configs/stage-nginx-config.yaml:19]: [detailed error messages](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/ca2169576e8777167bbdd96889e609d5e189c0af)

File: k8s/nginx-ingress/global-configs/prod-nginx-config.yaml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    error_page 403 = @access_denied;
+    error_page 401 = @unauthorized;"
+  server-snippets: "
+    location @internal_server_error { return 500 '{\"success\":false, \"message\":\"Something went wrong!\", \"error\":\"Internal Server Error\"}'; }
+    location @page_not_found { return 404 '{\"success\":false, \"message\":\"Resource not found :(\", \"error\":\"Not Found\"}'; }
+    location @access_denied { return 403 '{\"success\":false, \"message\":\"Access denied :|\", \"error\":\"Insufficient Permissions\"}'; }
+    location @unauthorized { return 401 '{\"success\":false, \"message\":\"Invalid authentication\", \"error\":\"Unauthorized.\"}'; }
+
+    location = /auth {
+      if ($request_uri !~ '/api/v') { return 200; } # All none api requests
```
Comment (123MwanjeMike) [inline k8s/nginx-ingress/global-configs/prod-nginx-config.yaml:22]: [don't authenticate any none api endpoints](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/95207b7020f5d63d0f3e05d1905d83ecf50efd1e)

**Comment by Baalmart** [comment]
> Next, we need the auth-service to automatically authorise;
> 
> * [x]  the endpoints that do not need authentication e.g login, register, forgot password, and home(/)
> * [ ]  other internal micro services and automatically renew the respective tokens.
> 
> cc @Baalmart

Thanks @123MwanjeMike  I think the logic should be handled at the load balancer level.
The auth service already does not authenticate those endpoints....so adding any more functionality around this is a bit unclear.

Happy to learn about your additional thoughts.

**Comment by 123MwanjeMike** [comment]
> > Next, we need the auth-service to automatically authorise;
> > 
> > * [x]   the endpoints that do not need authentication e.g login, register, forgot password, and home(/)
> > * [ ]   other internal micro services and automatically renew the respective tokens.
> > 
> > cc @Baalmart
> 
> Thanks @123MwanjeMike I think the logic should be handled at the load balancer level. The auth service already does not authenticate those endpoints....so adding any more functionality around this is a bit unclear.
> 
> Happy to learn about your additional thoughts.

Thanks @Baalmart.
We have also removed the authentication for those at the gateway level.
All is well now.

File: src/auth-service/utils/multitenancy.js [hunk 13]
```diff
@@ -1,29 +1,27 @@
 const mongodb = require("../config/dbConnection");
 const constants = require("../config/constants");
```
Comment (Baalmart) [inline src/auth-service/utils/multitenancy.js:2]: some of these changes are not expected in this PR, please revisit the way you resolved the merge conflicts. @123MwanjeMike

**Review by Baalmart — CHANGES_REQUESTED** [review]
Just left one example comment, looks like there is a need for us to revisit how we resolved the merge conflicts. @123MwanjeMike

File: .github/workflows/deploy-apis-to-production.yml [hunk 14]
```diff
@@ -541,7 +541,7 @@ jobs:
         run: |
           cd src/airflow/
           gcloud secrets versions access latest --secret="prod-env-airflow" > .env
-          kubectl create configmap --dry-run=client -o yaml --from-env-file=.env airflow-config | kubectl replace -f - -n production
+          kubectl create configmap --dry-run=client -o yaml --from-env-file=.env airflow-config | kubectl replace -f - -n pipeline
```
Comment (123MwanjeMike) [inline .github/workflows/deploy-apis-to-production.yml:544]: Correct airflow [pipeline](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/4caf7b18377f7dd8239b133399363f8fd08ed4ad) error

File: .prettierignore [hunk 15]
```diff
@@ -1,2 +1,2 @@
-k8s/nginx-ingress/global-configs/prod-nginx-config.yaml
-k8s/nginx-ingress/global-configs/stage-nginx-config.yaml
\ No newline at end of file
+k8s/nginx/production/global-config.yaml
+k8s/nginx/staging/global-config.yaml
```
Comment (123MwanjeMike) [inline .prettierignore:2]: [exclude configs from linter](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/a47f2c7af38cbc6dd8a695ebe9261663d352e0e1)

File: k8s/nginx/production/global-config.yaml [hunk 16]
```diff
@@ -4,30 +4,38 @@ metadata:
...
 
-    internal;
-    proxy_method            POST;
-    proxy_pass              http://airqo-auth-api-svc.production.svc.cluster.local:3000/api/v1/users/verify?tenant=airqo;
-    proxy_pass_request_body off;
-    proxy_set_header        Content-Length '''';
-    proxy_set_header        X-Original-URI $request_uri;
-    }'
+      # Mobile app specific exemptions. To be authenticated later
+      if ($request_uri ~ '/api/v1/view/measurements/app/insights') { return 200; }
```
Comment (123MwanjeMike) [inline k8s/nginx/production/global-config.yaml:31]: [mobile specific exepmtions](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/eeb86185fbe9b94b5c38abe54c1a36cd823f78fe) cc @NoahNsimbe

File: workflow-trigger [hunk 17]
```diff
@@ -1 +1 @@
-2022-09-23T20:00:00Z
+2022-10-20T20:00:00Z
```
Comment (123MwanjeMike) [inline workflow-trigger:1]: [trigger all](https://github.com/airqo-platform/AirQo-api/pull/1011/commits/1ba1505dd4c4caf131892e34ca5821f815d5c7ac)

File: k8s/nginx/staging/ingress.yaml [hunk 18]
```diff
@@ -194,13 +194,6 @@ spec:
                 name: airqo-stage-view-api-svc
                 port:
                   number: 8080
-          - path: /api/v1/views
```
Comment (123MwanjeMike) [inline k8s/nginx/staging/ingress.yaml:197]: Remove views path

**Review by 123MwanjeMike — COMMENTED** [review]
Guiding comments @Baalmart @NoahNsimbe

**Review by Baalmart — APPROVED** [review]
LGTM, I will leave the thorough review and the merge to be done by @NoahNsimbe 

Thanks @123MwanjeMike !

File: k8s/nginx/production/global-config.yaml [hunk 19]
```diff
@@ -4,30 +4,38 @@ metadata:
...
-    if ($request_uri ~ ''(/api/v1/users/loginUser)|(/api/v2/users/loginUser)'') { return 200; }
-    if ($request_uri ~ ''(/api/v1/users/forgotPassword)|(/api/v2/users/forgotPassword)'') { return 200; }
-    if ($request_uri ~ ''(/api/v1/users/candidates/register)|(/api/v2/users/candidates/register)'') { return 200; }
-    if ($request_uri !~ ''/api/v'') { return 200; } # All none api requests
+      if ($request_uri ~ '(/api/v1/users/emailAuth)|(/api/v2/users/emailAuth)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/loginUser)|(/api/v2/users/loginUser)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/emailLogin)|(/api/v2/users/emailLogin)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/forgotPassword)|(/api/v2/users/forgotPassword)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/candidates/register)|(/api/v2/users/candidates/register)') { return 200; }
+      if ($request_uri ~ '/airflow/api/v') { return 200; }
```
Comment (NoahNsimbe) [inline k8s/nginx/production/global-config.yaml:27]: Hi @123MwanjeMike, return 200 for all airflow requests. It has its own well designed authentication mechanism.
i.e `if ($request_uri ~ '/airflow') { return 200; }`

File: k8s/nginx/staging/global-config.yaml [hunk 20]
```diff
@@ -4,31 +4,39 @@ metadata:
...
+    location @unauthorized { return 401 '{\"success\":false, \"message\":\"Invalid authentication\", \"error\":\"Unauthorized.\"}'; }
+    
+    location = /auth {      
+      if ($request_method = 'OPTIONS') { return 204; } 
+      if ($request_uri ~ '(/api/v1/users/emailAuth)|(/api/v2/users/emailAuth)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/loginUser)|(/api/v2/users/loginUser)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/emailLogin)|(/api/v2/users/emailLogin)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/forgotPassword)|(/api/v2/users/forgotPassword)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/candidates/register)|(/api/v2/users/candidates/register)') { return 200; }
+      if ($request_uri ~ '/airflow/api/v') { return 200; }
```
Comment (NoahNsimbe) [inline k8s/nginx/staging/global-config.yaml:28]: Same here

Comment (123MwanjeMike) [reply k8s/nginx/staging/global-config.yaml:28]: That is already taken care of by `if ($request_uri !~ '/api/v') { return 200; } # All none api requests`

File: k8s/nginx/staging/global-config.yaml [hunk 21]
```diff
@@ -4,31 +4,39 @@ metadata:
...
+    
+    location = /auth {      
+      if ($request_method = 'OPTIONS') { return 204; } 
+      if ($request_uri ~ '(/api/v1/users/emailAuth)|(/api/v2/users/emailAuth)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/loginUser)|(/api/v2/users/loginUser)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/emailLogin)|(/api/v2/users/emailLogin)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/forgotPassword)|(/api/v2/users/forgotPassword)') { return 200; }
+      if ($request_uri ~ '(/api/v1/users/candidates/register)|(/api/v2/users/candidates/register)') { return 200; }
+      if ($request_uri ~ '/airflow/api/v') { return 200; }    
+      if ($request_uri !~ '/api/v') { return 200; } # All none api requests
```
Comment (123MwanjeMike) [inline k8s/nginx/staging/global-config.yaml:29]: here @NoahNsimbe

File: k8s/nginx/production/global-config.yaml (hunk 19 above)
Comment (123MwanjeMike) [reply k8s/nginx/production/global-config.yaml:27]: https://github.com/airqo-platform/AirQo-api/pull/1011#discussion_r1000264887

File: k8s/nginx/staging/global-config.yaml (hunk 21 above)
Comment (NoahNsimbe) [reply k8s/nginx/staging/global-config.yaml:29]: cool..thanks @123MwanjeMike

**Pull request merged**